ARROW-15901: [C++] Support custom output field names in Substrait

rtpsw opened the pull request:
Replaces https://github.com/apache/arrow/pull/12601

rtpsw commented:
@westonpace, I don't know why the errors in Travis CI occurred. Any idea how to resolve them?

vibhatha commented on cpp/src/arrow/compute/exec/options.h:
```diff
@@ -232,10 +232,13 @@ class ARROW_EXPORT SinkNodeConsumer {
+                                    std::vector<std::string> names = {})
+      : consumer(std::move(consumer)), names(std::move(names)) {}
 
   std::shared_ptr<SinkNodeConsumer> consumer;
+  /// \brief Names to rename the sink's schema fields to
+  std::vector<std::string> names;
```
Nit: May be distinguish the member variables with an underscore?

```c++
std::shared_ptr<SinkNodeConsumer> consumer_;
std::vector<std::string> names_;
```

And change the rest accordingly. Looking into `ConsumingSinkNode`, that is the style adopted.

westonpace replied:
In this case I think the naming scheme is correct.  There are two categories of objects we deal with, passive data objects and more complex stateful objects.  The google style guide likes to use structs for the former and classes for the latter and so it describes this as [structs vs. classes](https://google.github.io/styleguide/cppguide.html#Structs_vs._Classes).

In Arrow we are pretty inconsistent about applying `struct` and `class`.  Many of our "options" objects are `class` when they should be `struct` for example.  However, we are pretty consistent about the naming.  For a passive data object all properties should be public and [should not have trailing underscores](https://google.github.io/styleguide/cppguide.html#Variable_Names).

Options objects should always be passive data objects.  Things that travel back and forth across library boundaries should generally be either passive data objects or pure virtual interfaces when possible.  So I do not think a change is needed here.  This options object should be a passive data object and it is following the correct rules for a passive data object.

vibhatha replied:
Ah I see your point. Thank you for the detailed explanation. 👍

westonpace commented on cpp/src/arrow/compute/exec/options.h:
```diff
@@ -232,10 +232,13 @@ class ARROW_EXPORT SinkNodeConsumer {
+      : consumer(std::move(consumer)), names(std::move(names)) {}
 
   std::shared_ptr<SinkNodeConsumer> consumer;
+  /// \brief Names to rename the sink's schema fields to
```
```suggestion
  /// \brief Names to rename the sink's schema fields to
  ///
  /// If specified then names must be provided for all fields
```

westonpace commented on cpp/src/arrow/engine/substrait/serde.cc:
```diff
@@ -67,15 +67,19 @@ Result<std::vector<compute::Declaration>> DeserializePlan(
     auto sink_decl = compute::Declaration::Sequence({
         std::move(decl),
-        {"consuming_sink", compute::ConsumingSinkNodeOptions{consumer_factory()}},
+        {"consuming_sink", compute::ConsumingSinkNodeOptions{consumer_factory(), names}},
```
```suggestion
        {"consuming_sink", compute::ConsumingSinkNodeOptions{consumer_factory(), std::move(names)}},
```

westonpace reviewed (changes requested):
I have a few minor comments but let's also add a test case for this.

This also isn't yet a full implementation of the Substrait spec as it doesn't handle nested fields.  For example, if there are three columns and the first column is a struct column with two fields then the names vector should have 5 items in it.  However, if we want to postpone naming of nested fields to a future PR I think that is fine, just make sure to add a JIRA for it.

rtpsw commented:
> I have a few minor comments but let's also add a test case for this.

Added.

> This also isn't yet a full implementation of the Substrait spec as it doesn't handle nested fields. For example, if there are three columns and the first column is a struct column with two fields then the names vector should have 5 items in it. However, if we want to postpone naming of nested fields to a future PR I think that is fine, just make sure to add a JIRA for it.

I opted to postpone, and updated the title and the [ARROW-15901](https://issues.apache.org/jira/browse/ARROW-15901) accordingly.

rtpsw commented:
@westonpace, the AppVeyor failure seems unrelated to the change here. How to proceed?

westonpace commented on cpp/src/arrow/compute/exec/sink_node.cc:
```diff
@@ -263,18 +263,21 @@ class SinkNode : public ExecNode {
                  /*num_outputs=*/0),
-        consumer_(std::move(consumer)) {}
+        consumer_(std::move(consumer)),
+        names_(names) {}
```
```suggestion
        names_(std::move(names)) {}
```

westonpace reviewed (approved):
The appveyor failure seems unrelated.  One small suggestion and I think this is good to go.

rtpsw commented:
@westonpace, this has been pending for a while. How to proceed?

westonpace commented:
I think this is ready to go.  If @vibhatha has any more thoughts he can of course still mention them and we can address in a follow-up.
